staging/src/k8s.io/sample-controller/integration_test.go

```diff
@@ -0,0 +1,215 @@
+	}
+
+	// Create a test namespace
+	ns := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "namespace-for-test",
+		},
+	}
+	_, err = kubeClient.Core().Namespaces().Create(ns)
+	if err != nil {
+		t.Fatalf("error creating namespace: %v", err)
+	}
+
+	// XXX: Can't use the generated client to create Foo resources.
+	// It fails with:
+	//	integration_test.go:92: error creating example foo: object *v1alpha1.Foo does not implement the protobuf marshalling interface and cannot be encoded to a protobuf message
+	// replicas := int32(1)
+	// exampleFoo := &sampleapi.Foo{
+	//	ObjectMeta: metav1.ObjectMeta{
+	//		Name: "example-foo",
+	//	},
+	//	Spec: sampleapi.FooSpec{
+	//		DeploymentName: "example-foo",
+	//		Replicas:       &replicas,
+	//	},
+	// }
+	// _, err = exampleClient.SamplecontrollerV1alpha1().Foos(ns.Name).Create(exampleFoo)
+	// if err != nil {
+	//	t.Fatalf("error creating example foo: %v", err)
+	// }
+
+	// Create an instance of the CRD using dynamic client.
+	// Following example from https://github.com/kubernetes/apiextensions-apiserver/blob/release-1.8/test/integration/registration_test.go#L42
+	exampleFoo := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": "samplecontroller.k8s.io/v1alpha1",
+			"kind":       "Foo",
+			"metadata": map[string]interface{}{
+				"namespace": ns.Name,
+				"name":      "example-foo",
+			},
+			"spec": map[string]interface{}{
+				"deploymentName": "example-foo",
+				"replicas":       123,
+			},
+		},
+	}
+	resourceClient := dynamicClient.Resource(
+		&metav1.APIResource{
+			Name:       crd.Spec.Names.Plural,
+			Namespaced: crd.Spec.Scope != apiextensionsv1beta1.ClusterScoped,
+		},
+		ns.Name,
+	)
+	_, err = resourceClient.Create(exampleFoo)
+	if err != nil {
+		t.Fatalf("error creating example foo: %v", err)
+	}
+
+	// The example client can find the created Foo resource
+	found, err := exampleClient.SamplecontrollerV1alpha1().Foos(ns.Name).Get(
+		"example-foo",
+		metav1.GetOptions{},
+	)
+	if err != nil {
+		t.Fatalf("error getting example-foo: %v", err)
+	}
+	t.Logf("Found example-foo: %v", found)
+
+	// Create and start the sample-controller and its informers
+	kubeInformerFactory := kubeinformers.NewSharedInformerFactory(kubeClient, 0)
+	exampleInformerFactory := informers.NewSharedInformerFactory(exampleClient, 0)
+
+	controller := NewController(
+		kubeClient,
+		exampleClient,
+		kubeInformerFactory,
+		exampleInformerFactory,
+	)
+
+	stopCh := make(chan struct{})
+	controllerFinished := make(chan struct{})
+	defer func() {
+		close(stopCh)
+		<-controllerFinished
+	}()
+
+	kubeInformerFactory.Start(stopCh)
+	exampleInformerFactory.Start(stopCh)
+
+	go func() {
+		defer close(controllerFinished)
+		err := controller.Run(2, stopCh)
+		if err != nil {
+			t.Fatalf("error running controller: %v", err)
+		}
+	}()
+
+	// Wait 60s for the controller to create a deployment
+	err = wait.PollImmediate(
+		5*time.Second,
+		60*time.Second,
```

Comment:
Can these numbers be pulled out into const's to make it clearer what they are used for & so they can be reused throughout the package/tests? (e.g. `const longTimeout = 60 * time.second`)